packages/arb-ts/src/lib/assetBridger/tokenBridger.ts

```diff
@@ -0,0 +1,719 @@
+import { Provider, Filter } from '@ethersproject/abstract-provider'
+import { PayableOverrides } from '@ethersproject/contracts'
+import { Zero, MaxUint256 } from '@ethersproject/constants'
+import { Logger } from '@ethersproject/logger'
+import { BigNumber, ethers } from 'ethers'
+
+import {
+  L1GatewayRouter__factory,
+  L2GatewayRouter__factory,
+  L1ERC20Gateway__factory,
+  L1WethGateway__factory,
+  StandardArbERC20__factory,
+  L2ArbitrumGateway__factory,
+  L2ArbitrumGateway,
+  ERC20__factory,
+  L1GatewayRouter,
+  StandardArbERC20,
+  ERC20,
+} from '../abi'
+import { WithdrawalInitiatedEvent } from '../abi/L2ArbitrumGateway'
+import { GatewaySetEvent } from '../abi/L1GatewayRouter'
+import {
+  GasOverrides,
+  L1ToL2MessageGasEstimator,
+} from '../message/L1ToL2MessageGasEstimator'
+import { SignerProviderUtils } from '../utils/signerOrProvider'
+import { L2Network } from '../utils/networks'
+import { isError } from '../utils/lib'
+import { MultiCaller } from '../utils/multicall'
+import { ArbTsError, MissingProviderArbTsError } from '../errors'
+import { EventFetcher } from '../utils/eventFetcher'
+
+import { EthDepositBase, EthWithdrawParams } from './ethBridger'
+import { AssetBridger } from './assetBridger'
+import {
+  L1ContractTransaction,
+  L1TransactionReceipt,
+} from '../message/L1ToL2Message'
+import {
+  L2ContractTransaction,
+  L2TransactionReceipt,
+} from '../message/L2ToL1Message'
+
+export interface TokenApproveParams {
+  /**
+   * L1 signer whose tokens are being approved
+   */
+  l1Signer: Signer
+
+  /**
+   * L1 address of the ERC20 token contract
+   */
+  erc20L1Address: string
+
+  /**
+   * Amount to approve. Defaults to max int.
+   */
+  amount?: BigNumber
+
+  /**
+   * Transaction overrides
+   */
+  overrides?: PayableOverrides
+}
+
+export interface TokenDepositParams extends EthDepositBase {
+  /**
+   * L1 address of the token ERC20 contract
+   */
+  erc20L1Address: string
+
+  /**
+   * L2 address of the entity receiving the funds. Defaults to the l1FromAddress
+   */
+  destinationAddress?: string
+
+  /**
+   * Overrides for the retryable ticket parameters
+   */
+  retryableGasOverrides?: GasOverrides
+}
+
+export interface TokenWithdrawParams extends EthWithdrawParams {
+  /**
+   * L1 address of the token ERC20 contract
+   */
+  erc20l1Address: string
+}
+
+/**
+ * Bridger for moving ERC20 tokens back and forth betwen L1 to L2
+ */
+export class TokenBridger extends AssetBridger<
+  TokenDepositParams,
+  TokenWithdrawParams
+> {
+  public static MAX_APPROVAL = MaxUint256
+  public static MIN_CUSTOM_DEPOSIT_MAXGAS = BigNumber.from(275000)
+
+  public constructor(l2Network: L2Network) {
+    super(l2Network)
+  }
+
+  /**
+   * Get the address of the l1 gateway for this token
+   * @param erc20L1Address
+   * @param l1Provider
+   * @returns
+   */
+  public async getL1GatewayAddress(
+    erc20L1Address: string,
+    l1Provider: Provider
```

Comment:
Revisiting this general q of how to handle signers and providers: I'm probably generally overthinking this, but we should be consistent whatever we end up choosing; i.e., if we're gonna go with this pattern of requiring signers/providers to be passed in as params in each method, we should update L1ToL2MessageCreator accordingly (which currently sets up the signer in the constructor).  

One thing that feels slightly off to me is that we require a network object be passed into the constructor, but then ask for a signer/provider when calling a method, effectively assuming that the network & provider correspond. 

Something that might be useful (for me anyway) would be to see what a simple script that instantiates one of these classes (i.e., TokenApproveParams) looks like (without, i.e., using instantiateBridge) and see if there's some utility helpers we could include to it neaten up the flow.

Reply:
(See, i.e., https://github.com/OffchainLabs/arbitrum/pull/1907/files)

Reply:
as discussed, we'll revisit this, but can ship for beta in its current form 👍